### PR TITLE
Feat: create module for pretextmap

### LINKED ISSUE #1947: new module: PRETEXT/MAP
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by gertrudeln**
## Description

User request to add module for PretextMap. Added module alongside with test config.

## PR checklist

Closes #1947

- [x] This comment contains a description of changes (with reason).
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

@nf-core-bot fix linting

### CODE REVIEW

File: modules/nf-core/pretextmap/main.nf [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    tag "$meta.id"
+    label 'process_single'
+
+    conda (params.enable_conda ? "bioconda::pretextmap=0.1.9 bioconda::samtools=1.16.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/mulled-v2-f3591ce8609c7b3b33e5715333200aa5c163aa61:c6242a6c1a522137de7a9e9ff90779ede11cf5c5-0':
+        'quay.io/biocontainers/mulled-v2-f3591ce8609c7b3b33e5715333200aa5c163aa61:c6242a6c1a522137de7a9e9ff90779ede11cf5c5-0' }"
+
+    input:
+    tuple val(meta), path(bam)
```
Comment (priyanka-surana) [inline modules/nf-core/pretextmap/main.nf:12]: PretextMap also takes `pairs` and `sam` file. Call it `input` instead. And describe the different options `sam/bam` or `pairs` in `meta.yml`

File: modules/nf-core/pretextmap/main.nf [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    samtools \\
+        view \\
+        -h \\
+        $bam | PretextMap \\
+        $args \\
+        -o ${prefix}.pretext
```
Comment (priyanka-surana) [inline modules/nf-core/pretextmap/main.nf:30]: The commands are slightly different with `sam/bam` vs `pairs` input and both needs to be implemented.
https://github.com/wtsi-hpag/PretextMap

File: modules/nf-core/pretextmap/meta.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      doi: ""
+      licence: "['MIT']"
+
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - bam:
```
Comment (priyanka-surana) [inline modules/nf-core/pretextmap/meta.yml:20]: Call it `input` and describe the different options

File: tests/modules/nf-core/pretextmap/main.nf [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+workflow test_pretextmap {
+    
+    input = [
+        [ id:'test', single_end:false ], // meta map
+        file(params.test_data['sarscov2']['illumina']['test_paired_end_bam'], checkIfExists: true)
+    ]
+
+    PRETEXTMAP ( input )
+}
```
Comment (priyanka-surana) [inline tests/modules/nf-core/pretextmap/main.nf:15]: Include test with both `bam/sam` and `pairs`

**Review by priyanka-surana — CHANGES_REQUESTED** [review]

File: tests/modules/nf-core/pretextmap/main.nf (hunk 4 above)
Comment (marrip) [reply tests/modules/nf-core/pretextmap/main.nf:35]: I added tests for `bam/cram`. `sam` should not be tested according to James Fellows Yates.

File: modules/nf-core/pretextmap/meta.yml (hunk 3 above)
Comment (marrip) [reply modules/nf-core/pretextmap/meta.yml]: done

File: modules/nf-core/pretextmap/main.nf (hunk 2 above)
Comment (marrip) [reply modules/nf-core/pretextmap/main.nf]: Added support for that but somehow it broke my module. Not sure what the problem is as I don't get any informative logs...

File: modules/nf-core/pretextmap/main.nf (hunk 1 above)
Comment (marrip) [reply modules/nf-core/pretextmap/main.nf]: done

**Comment by marrip** [comment]
I don't get these tests to work. I can't make any sense of the `pytest` output. Any hints or advice would be highly appreciated.

**Comment by marrip** [comment]
@nf-core/modules-team

**Comment by jfy133** [comment]
@marrip in this case you need to look at the actual nextflow logs (pytest traceback sucks, I hate it personally).

It appears your conda environment is not building correctly (as far as I can tell).If you scroll to the bottom of the summary page of the CI tests, you can find the artefact containing all the logs:

https://github.com/nf-core/modules/actions/runs/3241811168

In which you should see:

```
Caused by:
  Process `test_pretextmap_bam:PRETEXTMAP (test)` terminated with an error exit status (127)

Command executed:

  if [[ test.paired_end.sorted.bam == *.pairs.gz ]]; then
      zcat test.paired_end.sorted.bam | PretextMap \
           \
          -o test.pretext
  else
      samtools \
          view \
          -h \
          test.paired_end.sorted.bam | PretextMap \
           \
          -o test.pretext
  fi
  
  cat <<-END_VERSIONS > versions.yml
  "test_pretextmap_bam:PRETEXTMAP":
      pretextmap: $(PretextMap | grep "Version" | sed 's/PretextMap Version //g')
      samtools: $(echo $(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*$//' ))
  END_VERSIONS

Command exit status:
  127

Command output:
  (empty)

Command error:
  .command.sh: line 12: PretextMap: command not found
```

Areyou able to test with conda locally?

**Comment by marrip** [comment]
Hey @jfy133 ! Thanks for your reply. I checked testing it locally and it fails as well, looking into the logs, I see that indeed the conda environment did not build correctly using `samtools=1.16.1`:
```
Package libgcc-ng conflicts for:
samtools=1.16.1 -> ncurses[version='>=6.3,<7.0a0'] -> libgcc-ng[version='>=11.2.0|>=7.5.0']
pretextmap=0.1.9 -> libgcc-ng[version='>=10.3.0|>=9.4.0']
samtools=1.16.1 -> libgcc-ng[version='>=12']
```
I tried to with `samtools=1.15.1`:
```
Package _libgcc_mutex conflicts for:
samtools=1.15.1 -> libgcc-ng[version='>=10.3.0'] -> _libgcc_mutex[version='*|0.1',build=main]
pretextmap=0.1.9 -> libgcc-ng[version='>=10.3.0'] -> _libgcc_mutex[version='*|0.1',build=main]
```
which is extremely confusing as both tools request the same version of `libgcc-ng`. Nevertheless, I was able to run these tests successfully before I changed the `script` - the conda environment was the same as now. It seems they also fail with both docker and singularity which is using a mulled container which contains both tools - I checked locally.
And I agree, the pytest output is more so confusing than it is helpful. I would expect some indication of the source of the error...I don't really know how to trouble shoot this. I also checked the nextflow logs but for my test with docker it did not say anything helpful either...

**Comment by jfy133** [comment]
> Hey @jfy133 ! Thanks for your reply. I checked testing it locally and it fails as well, looking into the logs, I see that indeed the conda environment did not build correctly using `samtools=1.16.1`:
> 
> ```
> Package libgcc-ng conflicts for:
> samtools=1.16.1 -> ncurses[version='>=6.3,<7.0a0'] -> libgcc-ng[version='>=11.2.0|>=7.5.0']
> pretextmap=0.1.9 -> libgcc-ng[version='>=10.3.0|>=9.4.0']
> samtools=1.16.1 -> libgcc-ng[version='>=12']
> ```
> 
> I tried to with `samtools=1.15.1`:
> 
> ```
> Package _libgcc_mutex conflicts for:
> samtools=1.15.1 -> libgcc-ng[version='>=10.3.0'] -> _libgcc_mutex[version='*|0.1',build=main]
> pretextmap=0.1.9 -> libgcc-ng[version='>=10.3.0'] -> _libgcc_mutex[version='*|0.1',build=main]
> ```
> 
> which is extremely confusing as both tools request the same version of `libgcc-ng`. Nevertheless, I was able to run these tests successfully before I changed the `script` - the conda environment was the same as now. It seems they also fail with both docker and singularity which is using a mulled container which contains both tools - I checked locally. And I agree, the pytest output is more so confusing than it is helpful. I would expect some indication of the source of the error...I don't really know how to trouble shoot this. I also checked the nextflow logs but for my test with docker it did not say anything helpful either...

Is that after installing in an isolated environment `conda create -n pretextmap -c bioconda pretext=1.<...> samtools=<...>`?

**Comment by marrip** [comment]
> Is that after installing in an isolated environment `conda create -n pretextmap -c bioconda pretext=1.<...> samtools=<...>`?
yes

**Comment by marrip** [comment]
@priyanka-surana I made the requested changes. Could you please check again? Thank you.

**Review by priyanka-surana — APPROVED** [review]
LGTM 👍

**Pull request merged**